transactions/read-only/parallel.md

```diff
@@ -0,0 +1,214 @@
+- `read-only-write-window-time-us`: time in microseconds the `write` window lasts. For this option to take effect, `read-only-transaction-threads` must be set to greater than 0. Default to `200,000` microseconds. 
+- `read-only-read-window-time-us`: time in microseconds the `read` window lasts. For this option to take effect, `read-only-transaction-threads` must be set to greater than 0. Default to `60,000` microseconds.
+
+Note: `read-only-max-transaction-ms` (time in milliseconds a read-only transaction can execute before being considered invalid) was considered. But to keep new options minimal, for now, use the existing `max-transaction-time` for read-only transactions.
+
+### Data Structures 
+
+Three queues are proposed: the existing internal priority queue in `execution_priority_queue` of appbase is replaced with read operation queue and write operation queue, and a new read-only transaction queue is added.
+- `read operation queue` stores read operations safe to read-only transaction execution. Those operations are executed in both `read window` and `write window`. `appbase`'s `post` method is enhanced to have a new parameter indicating operation type.
+- `write operation queue` stores write operations not safe to read-only transaction execution. Those operations are only executed in `write` window.
+- `read-only transaction queue` stores read-only transactions. It is a deque.
+
+### Window Processing
+
+In `write window`, operations in `write operation queue` and `read operation queue` are executed by the main thread, while read-only transactions are queued in the read-only transaction queue for later execution in the `read window`. To select next operation to execute, priorities of the front operation in `write operation queue` and the front operation in `read operation queue` are compared; the one with higher priority is chosen. If the priorities are tied, the sequence numbers (`order` field in the existing code base) are compared; the one scheduled earlier is selected. This is the existing behavior.
+
+In `read window`, only operations in `read operation queue` are executed by the main thread, and transactions in read-only queue are executed by the read-only. Design details are:
+- At the start of `read-window`, all threads in the pool are signaled to start process read-only transactions. Each thread enters a loop of pulling one transaction a time from the front of the read-only transaction queue and executing it. New incoming read-only transactions are placed at the back of the queue. The queue is protected by mutex.
+- If less than 5% (hardcoded for now) of `read-only-read-window-time-us` time remains in the `read window`, a thread will stop scheduling new transaction for execution and exit the execution loop.
+- At the end of the window, non-finished transactions are put to the front of the read-only transaction queue so that they can be executed the first in the next round.
```

Comment:
We talked about this. As they are all read-only transactions, as long as we process the unfinished ones from last round before newer ones, it might be OK not to preserve the order among unfinished ones. That's why I mentioned using deque for simpler implementation. But I can change to use sequence number to preserve the order.